### ISSUE TITLE
How many experts are activated for each sub-task of Domainbed benchmarks?

### ISSUE THREAD

**Issue by psmiz** [LINs-lab/DynMoE#5]
Thank you for providing source code for DynMoE.
I have question regarding the use of DynMoE on Vision task.
Can I get to know the exact number of activated experts (top-k) corresponding to each of the sub-task of Domainbed benchmarks (used in Table 1)?

**Comment by QAQdev** [comment]
Thanks for your attention on DynMoE!
We regret that we do not have top-k statistics available for DomainBed😔
But if you really want to know it, I suggest you to do the following change:

1. Normally during training, DynMoE collects routing records for a while and updates the experts. During inference, we do not collect routing records. **So if you want to get the records, the first thing to do is to enable `begin_record_routing()` once at the very beginning at inference, and disable `end_record_routing()` or `adaptive_update_experts()`**. This will collect all records during inference for later processing.

**See the following codes, it controls the update logic. Just comment these lines will only enable `begin_record_routing()`.**

https://github.com/LINs-lab/DynMoE/blob/0a3109131559b30669f474df8f921df13a5d2c11/EMoE/Vision/domainbed/algorithms.py#L406-L418

2. See the routing method below, you can define a variable `total_tokens` and add `total_tokens += logits.shape[0]`. **This calculates total tokens processed during inference.**

https://github.com/LINs-lab/DynMoE/blob/0a3109131559b30669f474df8f921df13a5d2c11/EMoE/tutel/tutel/impls/moe_layer.py#L367-L368

3. **The following codes maintain the routing records, i.e. what tokens routed to what experts.** You could simply save the `self.routing_records` variable to get total records.

https://github.com/LINs-lab/DynMoE/blob/0a3109131559b30669f474df8f921df13a5d2c11/EMoE/tutel/tutel/impls/moe_layer.py#L445-L460

4. Now run on DomainBed for once and that's it!🤪

5. After you get total tokens and routing records during inference, you could **calculate top-k by routing records / total tokens**.

**Comment by psmiz** [comment]
Thanks for quick response! I'll try to follow the steps you suggested.
